composer/trainer/trainer.py

```diff
@@ -1488,23 +1484,36 @@ def _get_autoresume_checkpoint(
             dist.broadcast_object_list(save_latest_remote_file_name_list, src=0)
             save_latest_remote_file_name = save_latest_remote_file_name_list[0]
 
-            # download the checkpoint on local rank 0 of all nodes
+            # try to download the checkpoint on local rank 0 of all nodes
             if dist.get_local_rank() == 0 and not os.path.exists(latest_checkpoint_path):
                 log.debug(f'Attempting to download the checkpoint {save_latest_remote_file_name} on to all nodes')
                 os.makedirs(save_folder, exist_ok=True)
                 self._try_checkpoint_download(latest_checkpoint_path, save_latest_remote_file_name, loggers,
                                               load_progress_bar)
-            dist.barrier()
-            # At this point the rank 0 filepath should exist on all ranks
-            latest_checkpoint_exists_on_all_ranks = self.state.device.tensor_to_device(
-                torch.tensor([os.path.exists(latest_checkpoint_path)], dtype=torch.uint8))
-            dist.all_reduce(latest_checkpoint_exists_on_all_ranks, reduce_operation='MIN')
+
+            signal_file_path = os.path.join(os.path.dirname(latest_checkpoint_path),
+                                            '.local_rank0_completed_autoresume')
+            if dist.get_local_rank() == 0:
+                with open(signal_file_path, 'wb') as f:
+                    f.write(b'local_rank0_completed_autoresume')
+
+            # avoid the collective call until the local rank zero has finished trying to download the checkpoint
+            # so that we don't timeout for large downloads
+            with dist.local_rank_zero_download_and_wait(signal_file_path):
+                dist.barrier()
```

Comment:
The order here is a bit confusing because there's a context and a barrier. Can we split them out and make it more explicit what happens in what order?

Reply:
so you mean
```
with dist.local_rank_zero_download_and_wait(signal_file_path):
    pass
dist.barrier()
```
?

Reply:
Yea maybe? I think that might help it be more explicit?

Reply:
given that I can't test multinode again right now, id like to merge it as is. i still have a jira to clean this section up after your refactor gets in. is that ok?

Reply:
sure, lets go! 🚀